### PR TITLE
Implement `drive_id` class more fully

### LINKED ISSUE #93: Apply drive_id class to the id variable in a dribble
It just occurred to me that the `id` variable in a dribble could have type `drive_id`. We already have this class and we can apply it inside the functions we already have for creating dribbles. This makes it more attractive to write a format method for `drive_id` objects that truncates / ellipsizes the ids. This is basically a revival of #5.

### PR DESCRIPTION
**Pull request by jennybc**
Closes #93 

*Some aspects of this PR were done in a certain way, in order to generate discussion for a live code review in a group meeting*

---

googledrive has always had `drive_id`, a lightweight S3 class to mark a string (or character vector) as file ID, instead of a filename.

`drive_id` pre-dated vctrs and was implemented in a very minimal way.

Overall, it’s been OK, but there are annoying downsides:

-   File IDs are almost meaningless to humans, but they take up lots of valuable
    space when printing `dribble`s, often at the expense of `name`, which is
    often more interesting to the user.
-   Any sort of manipulation tends to lose the `drive_id` class.
-   You can put strings in a `drive_id` object that are definitely not valid, like
    `""`.

Here’s what I mean:

``` r
drive_find(n_max = 3)
#> # A dribble: 3 x 3
#>   name                        id                                   drive_resource  
#> * <chr>                       <chr>                                <list>          
#> 1 foo_sheet-TEST-drive_publi… 1CEefQCUc5T7B4yrawnNfqwdWEbiDyDs9E5… <named list [36…
#> 2 foo_doc-TEST-drive_publish  1oUrQNg-2lcAieZyCqoQ_vDwYLMVzhN4-oO… <named list [36…
#> 3 imdb_latin1_csv-TEST-drive… 1o_7f6WRXN-CxgijwgfpDiust8A3zKcNI    <named list [39…

(ids <- as_id(month.abb))
#>  [1] "Jan" "Feb" "Mar" "Apr" "May" "Jun" "Jul" "Aug" "Sep" "Oct" "Nov" "Dec"
#> attr(,"class")
#> [1] "drive_id"
just_two <- ids[c(2, 5)]
just_two
#> [1] "Feb" "May"
class(just_two)
#> [1] "character"

as_id(c("", "lskdjflslsei"))
#> [1] ""             "lskdjflslsei"
#> attr(,"class")
#> [1] "drive_id"
```

This PR implements `drive_id` more fully and uses the capabilities in vctrs and pillar to address the above.
Resources consulted:

-   [S3 chapter of Advanced R](https://adv-r.hadley.nz/s3.html)
-   [S3 vectors vignette in vctrs](https://vctrs.r-lib.org/articles/s3-vector.html)
-   [Printing vectors nicely in tibbles vignette in vctrs](https://vctrs.r-lib.org/articles/pillar.html)

Another relevant resource is [Davis’s notes](https://github.com/DavisVaughan/2020-06-01_dplyr-vctrs-compat) on
“adding dplyr and vctrs support to your tibble subclass so that it works with the tidyverse”.
I recently updated the technology behind the `dribble` subclass and, since the `id` column of a `dribble` should be a `drive_id` (but previously was not), these two efforts intersect.

TL;DR here’s how things work now:

*The final `drive_id` truncation strategy ended up being different. There is less truncation than shown here.*

``` r
dat <- drive_find("jenny", n_max = 5)
dat
#> # A dribble: 5 x 3
#>   name                                         id       drive_resource   
#>   <chr>                                        <drv_id> <list>           
#> 1 DESCRIPTION-TEST-drive_upload-jenny-8fb0b018 19TUFZL… <named list [40]>
#> 2 DESC-TEST-drive_mv-jenny-9db391f6            1YbwaJe… <named list [40]>
#> 3 DESC-TEST-drive_mv-jenny-4be97296            18wab1k… <named list [40]>
#> 4 DESC-TEST-drive_mv-jenny-4be97296            1IH6z0z… <named list [40]>
#> 5 DESC-TEST-drive_mv-jenny-b4c90052            1qQMpEJ… <named list [40]>

(ids <- as_id(month.abb))
#> <drive_id[12]>
#>  [1] Jan Feb Mar Apr May Jun Jul Aug Sep Oct Nov Dec
just_two <- ids[c(2, 5)]
just_two
#> <drive_id[2]>
#> [1] Feb May
class(just_two)
#> [1] "drive_id"   "vctrs_vctr"

as_id(c("", "lskdjflslsei"))
#> Error: A <drive_id> must match this regular expression: `^[a-zA-Z0-9_-]+$`
#> Invalid input:
#> x '""'
```

## Problems and observations

The `pillar::pillar_shaft()` method isn’t working how I want.
Current work-around is to always truncate the `drive_id`s in a tibble.
This is not great, because sometimes you really do need access to the whole id for copy/paste. <https://github.com/r-lib/pillar/issues/331>

*That 👆 was resolved by settling for the truncation normally done on `character` columns. It's not ideal but it's better than nothing.*

~Lots of `as.character(id)` is necessary when incorporating `drive_id`s into requests, because of:~

``` r
ids
#> <drive_id[12]>
#>  [1] Jan Feb Mar Apr May Jun Jul Aug Sep Oct Nov Dec
jsonlite::toJSON(ids)
#> Error: No method asJSON S3 class: vctrs_vctr
```

*That 👆 was resolved by inheriting base class (character) in `new_drive_id()`.*

<sup>Created on 2021-06-28 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0.9000)</sup>

### CODE REVIEW

File: R/compat-dplyr.R [hunk 1]
```diff
@@ -14,13 +14,13 @@ dribble_is_reconstructable <- function(data) {
   is_list(data) &&
     has_dribble_cols(data) &&
     has_dribble_coltypes(data) &&
+    id_can_be_drive_id(data$id) &&
```
Comment (jennybc) [inline R/compat-dplyr.R:17]: When deciding if a `dribble` can be reconstructed, now we need to also consider if the putative `id` column can be cast to `drive_id`.

File: R/dribble.R [hunk 2]
```diff
@@ -183,8 +189,8 @@ has_drive_resource <- function(x) {
 #' drive_find("alfa") %>% drive_rm()
 as_dribble <- function(x, ...) UseMethod("as_dribble")
 
-#' @export
-as_dribble.dribble <- function(x, ...) x
+# #' @export
+#as_dribble.dribble <- function(x, ...) x
```
Comment (jennybc) [inline R/dribble.R:193]: I think I'll just send `dribble`s through the `data.frame` method.

File: R/drive_cp.R [hunk 3]
```diff
@@ -96,13 +96,13 @@ drive_cp <- function(file,
   # load (path, name) into params
   if (!is.null(path)) {
     path <- as_parent(path)
-    params[["parents"]] <- list(path$id)
+    params[["parents"]] <- list(as.character(path$id))
```
Comment (jennybc) [inline R/drive_cp.R:99]: There's lots of this explicit `as.character()` business now, to appease jsonlite.

```{r eval = FALSE}
ids
#> <drive_id[12]>
#>  [1] Jan Feb Mar Apr May Jun Jul Aug Sep Oct Nov Dec
jsonlite::toJSON(ids)
#> Error: No method asJSON S3 class: vctrs_vctr
```

File: R/drive_id-class.R [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+  pillar::new_pillar_shaft_simple(out, min_width = 3)
+
+  # full <- format(x)
+  # trunc <- format(paste0(substr(x, 1, 7), cli::symbol$continue))
+  # pillar::new_pillar_shaft(
+  #   list(full = full, trunc = trunc),
+  #   width = pillar::get_max_extent(full),
+  #   min_width = pillar::get_max_extent(trunc),
+  #   class = "pillar_shaft_drive_id"
+  # )
```
Comment (jennybc) [inline R/drive_id-class.R:85]: I wish this worked but it does not quite do so (yet?).

File: R/drive_cp.R (hunk 3 above)
Comment (hadley) [reply R/drive_cp.R:99]: Can you provide a method for `asJSON`? Or maybe this is just fixed by adding the `character` class? Either way, we should talk to Jeroen and see if there's a jsonlite tweak to be made.

File: R/drive_id-class.R [hunk 5]
```diff
@@ -0,0 +1,162 @@
+new_drive_id <- function(x = character()) {
+  vec_assert(x, character())
+  new_vctr(x, class = "drive_id")
```
Comment (hadley) [inline R/drive_id-class.R:3]:
```suggestion
  new_vctr(x, class = c("drive_id", "character"))
```

Comment (hadley) [reply R/drive_id-class.R:3]: Or maybe 

```R
new_vctr(x, class = "drive_id", inherit_base_type = TRUE)
```

File: R/drive_id-class.R [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+validate_drive_id <- function(x) {
+  ok <- is_valid_drive_id(x)
+  if (all(ok)) {
+    return(x)
+  }
+
+  # proceed with plain character vector
+  x <- unclass(x)
+  # pragmatism re: how to cli-style a path that is the empty string
+  x[!nzchar(x)] <- "\"\""
```
Comment (hadley) [inline R/drive_id-class.R:15]: File cli issue?

File: R/drive_id-class.R [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+is_drive_id <- function(x) {
+  inherits(x, "drive_id")
+}
+
+#' @export
+gargle_map_cli.drive_id <- function(x, ...) {
+  # x <- unclass(x)
+  # NextMethod()
+  # TODO: I know this is gross, but I cannot figure out any "nice" way
```
Comment (hadley) [inline R/drive_id-class.R:42]: Re-review after adding `character` to class; use `sloop::s3_dispatch()` to diagnose.

**Review by hadley — APPROVED** [review]

File: R/drive_id-class.R (hunk 6 above)
Comment (jennybc) [reply R/drive_id-class.R:15]: This is related to how I am using DIY methods to map cli style over a vector, with a pre-interpolation step.

I think when https://github.com/r-lib/cli/issues/265 gets sorted out, this problem basically goes away (or, at least, its nature would be very different).

File: R/drive_id-class.R (hunk 7 above)
Comment (jennybc) [reply R/drive_id-class.R]: Using `inherit_base_class = TRUE` (and thereby adding `character` to the class vector "fixed" this. Non-gross `NextMethod()` works now.
